File: frontend/admin/src/models/challenge.ts
import {Track} from './track';
import {Flag} from '@/models/flag';

/**
 * Challenge model
 */
export class Challenge {
  public id: number;
  public name: string;
  public description: string;
  public points?: number;
  public hidden: boolean;
  public flags?: Flag[];
  public category?: Track;
  public files?: File[];
  public links?: string[];
  public tags?: string[];

  constructor() {
    this.id = -1;
    this.name = '';
    this.description = '';
    this.hidden = true;
    this.files = [];
    this.links = [];
    this.tags = [];
    this.flags = [new Flag()];
  }
}

File: backend/JDISCTF/schemas/admin/challenge.py
"""Challenge marshmallow schemas"""
from flask_rebar import RequestSchema
from marshmallow import fields, Schema

from JDISCTF.schemas.admin.category import AdminCategorySchema
from JDISCTF.schemas.admin.flag import FlagSchema


class AdminChallengeRequestSchema(RequestSchema):
    """Request schema for creating a challenge"""
    category_id = fields.Integer(required=True)

    name = fields.String(required=True)
    description = fields.String(required=True)
    points = fields.Integer(required=True)
    flags = fields.Nested(FlagSchema, required=True, many=True)

    hidden = fields.Boolean(required=True)


class AdminChallengeListSchema(Schema):
    """Response schema for challenge listing"""

    id = fields.Integer(required=True)
    name = fields.String(required=True)
    category = fields.Nested(AdminCategorySchema, only="name", required=True)
    hidden = fields.Boolean(required=True)
    points = fields.Integer(required=True)


class AdminChallengeSchema(Schema):
    """Response schema for getting a challenge"""

    id = fields.Integer(required=True)
    category = fields.Nested(AdminCategorySchema)
    flags = fields.Nested(FlagSchema, many=True)

    name = fields.String(required=True)
    description = fields.String(required=True)
    points = fields.Integer(required=True)

    is_solved = fields.Boolean()
    hidden = fields.Boolean(required=True)
